### PR TITLE
Clean up filter code and add new options

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This PR does two things: clean up the filter code and add new options in an attempt to implement #435.

#435 is implemented by adding a select all/select none/invert selection menu to each filter list. Yes, the UI is clumsy, but this is simply something I'm bad at.

While trying to implement these changes, it occurred to me that the filter code is pretty ... weird. Functions were generated by macros which did all the same things and a class was split in two for no apparent reason. So I tried to clean this up without changing too much. This made the final implementation much easier.

The menu text "Unselect all" sounds strange to my ears, but this text already existed, so I reused it.

Testing and comments very welcome.

### Changes made:

1) Factor out common code of the *Filter classes into a common FilterBase subclass (thus mirroring the FilterModelBase class).
2) Fold MultiFilterInterface into FilterModelBase. There was no apparent reason for splitting the model into two classes.
3) Move FilterModel functions to FilterModelBase. Thus, remove the macro madness.
4) Add select all and invert selection functions to FilterModelBase and connect them to a popup-menu.

### CODE REVIEW

**Comment by janmulder** [comment]
Tested this PR:

- the UI change is good enough for me, and not sure how it can be improved (significantly)
- read the code changes, and lgtm
- but found a filter bug while testing this, but verified that that bug is also in current master.

**Comment by janmulder** [comment]
And the bug is: filter on 2 buddies. In my case, it says in the top "filtered 4 of 991 dives", but shown are only 2 dives in the dive list. Further, when switching the buddies off/on, suddenly all 4 dives appear. Not sure how well this problem reproduces. Maybe a new issue?

**Comment by bstoeger** [comment]
Thanks for testing!

Unfortunately I can't reproduce your issue. Sounds like the filter is working (correctly lists the number of dives), but for some reason the dive list is not updated? Could you send me an XML that shows the problem?

**Comment by janmulder** [comment]
And forget about the filter bug I thought I had found ... Is was user-error. Simply confused. Reason: the dives are partially in trips, and partially not in trips. What confused me, was that when selecting buddies off/on, sometimes trips are expanded, and sometimes not. So I was looking at one non-expanded trip, not realizing this.

File: desktop-widgets/simplewidgets.cpp [hunk 1]
```diff
@@ -500,109 +500,52 @@ void DiveComponentSelection::buttonClicked(QAbstractButton *button)
 	}
 }
 
-TagFilter::TagFilter(QWidget *parent) : QWidget(parent)
+FilterBase::FilterBase(FilterModelBase *model_, QWidget *parent)
+	: QWidget(parent)
+	, model(model_)
```
Comment (dirkhh) [inline desktop-widgets/simplewidgets.cpp:505]: we prefer the comma at the end of the previous line and the parent class on the declaration line:
```
FilterBase::FilterBase(FilterModelBase *model_, QWidget **parent) : QWidget(parent),
        model(model_)
```
and then additional initializers also add the comma in the line before.
I realize that the comma at the beginning style gives smaller diffs when you change things - I just have never seen a natural language where the comma moves to a new line in doing line breaks, so it bothers my brain.

I also am not a huge fan of ```model(model_)``` as that's too easy to confuse. How about ```model(modelParameter)```

File: qt-models/filtermodels.cpp [hunk 2]
```diff
@@ -79,6 +79,7 @@ CREATE_COMMON_METHODS_FOR_FILTER(SuitsFilterModel, count_dives_with_suit)
 CREATE_INSTANCE_METHOD(MultiFilterSortModel)
 
 FilterModelBase::FilterModelBase(QObject *parent) : QStringListModel(parent)
+						  , anyChecked(false)
```
Comment (dirkhh) [inline qt-models/filtermodels.cpp:29]: see previous comment - please move the comma behind the ```QStringListModel(parent)```

**Review by dirkhh — APPROVED** [review]
This seems like a very nice cleanup of code that I always struggled with. I would love to see a comment from @tcanabrava but he hasn't had much time for Subsurface in the last few months, so I think we can take this.
Given that it's Christmas Eve in Germany already (Frohe Weihnachten), I think I'll even take it with the little whitespace issues and simply fix them up in my own commit a little later.

Thanks

**Comment by dirkhh** [comment]
Thanks @janmulder for the testing. Sometimes we confuse ourselves and I love that you documented what happened.

**Pull request merged**